### ISSUE TITLE
[JS] False Negative : Unsafe Html Construction 

### ISSUE THREAD

**Issue by Naman-ntc** [github/codeql#8274]
In the following snippet, only the middle snippet is returned as vulnerable (the first and third calls are not marked vulnerable).

For the first case, I assume that _escaping_ single or double quotes will make the code safe. But in the third code snippet, it only _escapes_ `"` and should still be marked vulnerable? It seems that 
* the query only detects `attrVal.indexOf("\"") === -1` as a guard (and does not even consider `attrVal.indexOf("'") === -1` as a guard)
* does not use flow labels to check if both guards exist

```js
module.exports.guards = function(attrVal) {
    document.querySelector("#id").innerHTML = "<img alt=\"" + attrVal + "\"/>"; // NOT OK
    document.querySelector("#id").innerHTML = "<img alt=\"" + attrVal.replace(/"|'/g, "") + "\"/>"; // OK
    if (attrVal.indexOf("\"") === -1 && attrVal.indexOf("'") === -1) {
        document.querySelector("#id").innerHTML = "<img alt=\"" + attrVal + "\"/>"; // OK
    }
    if (attrVal.indexOf("'") === -1) {
        document.querySelector("#id").innerHTML = "<img alt=\"" + attrVal + "\"/>"; // NOT OK
    }
    if (attrVal.indexOf("\"") === -1) {
        document.querySelector("#id").innerHTML = "<img alt=\"" + attrVal + "\"/>"; // NOT OK
    }
}

```

**Comment by Naman-ntc** [comment]
Oh, sorry, I edited my example. Specifically, CodeQL produces an alert only on line 6 and not on lines 3 or 9. 
The example on line 3 along with the `// OK` is borrowed from `query-tests` and the other two are extended from it (with `// NOT OK` comment on what I felt output should be)
```js
module.exports.guards = function(attrVal) {
    if (attrVal.indexOf("\"") === -1 && attrVal.indexOf("'") === -1) {
        document.querySelector("#id").innerHTML = "<img alt=\"" + attrVal + "\"/>"; // OK [not reported vulnerable]
    }
    if (attrVal.indexOf("'") === -1) {
        document.querySelector("#id").innerHTML = "<img alt=\"" + attrVal + "\"/>"; // NOT OK [reported vulnerable]
    }
    if (attrVal.indexOf("\"") === -1) {
        document.querySelector("#id").innerHTML = "<img alt=\"" + attrVal + "\"/>"; // NOT OK [not reported vulnerable] [False Negative?]
    }
}

```

**Comment by hmakholm** [comment]
Thank you.

Can you elaborate further about what is bad about the last line? In my understanding, since the attribute `alt="...."` is explicitly constructed to be delimited with _double_ quotes, an embedded _single_ quote in `attrVal` would not help an attacker who controls `attrVal` escape the attribute-string context.

**Comment by hmakholm** [comment]
Hmm yes, that does look curious.
To help finding the right people to look at this, can you share the query ID or `.ql` filename of the query that produces the wrong alerts?

Then perhaps @github/codeql-javascript can take a look.

**Comment by Naman-ntc** [comment]
Yes, the vulnerability is https://codeql.github.com/codeql-query-help/javascript/js-html-constructed-from-input/

**Comment by erik-krogh** [comment]
That query don't really distinguish between HTML attributes that use `'` or `"`, and that causes the false-positive and false-negative in your last example.   
Specifically: We basically assume that single quotes don't exist, and just look for sanitizers that sanitize double-quotes.   
([See `QuoteGuard` in `Xss.qll`](https://github.com/github/codeql/blob/94cb5c2be4826f322e2cd7ecee800922730e7c92/javascript/ql/lib/semmle/javascript/security/dataflow/Xss.qll#L71-L85)). 

For now this is a deliberate limitation as doing it right would require tracking which quotes have been sanitized (using flow-labels). This would cause worse performance and a more complex implementation, and the benefit is quite limited.     

Btw. that sanitizer is used in `Xss.ql`, `StoredXss.ql`, `ReflectedXss.ql`, `XssThroughDom.ql`, `UnsafeJQueryPlugin.ql`, and `UnsafeHtmlConstruction.ql`.  
So fixing the issue would require changing all of those queries.  

--- 

Another reason for not fixing the issue is that we have the `IncompleteHtmlAttributeSanitization.ql` query that specifically looks for problems related to attribute sanitization.  
You can e.g. look at [this test case](https://github.com/github/codeql/blob/94cb5c2be4826f322e2cd7ecee800922730e7c92/javascript/ql/test/query-tests/Security/CWE-116/IncompleteSanitization/tst.js#L242-L257).
